README.md
# evince
Sandboxed Evince Document Viewer

This is a sandboxed version of the evince Document Viewer using seccomp filter (libseccomp)

The original application can be found here: https://wiki.gnome.org/Apps/Evince


Todo
----

- fix/clean makefiles
- prevent evince to try to open external links (causes crash)
- restructure evince process handling to further restrict seccomp filter


Goal
----

Before the application reads/renders the target document file, the list of allowed systemcalls should be near zero.


ChangeLog
commit ebe46b8a01fc85d127162e52699e16216cf27f7a
Author: Bastien Nocera <hadess@hadess.net>
Date:   2017-03-17

    shell: Fix incorrect return type

    Should be a gboolean, not a pointer.

    https://bugzilla.gnome.org/show_bug.cgi?id=780206

M	shell/ev-bookmarks.c

commit 90d023d5bdba0f3768925cbc363d89b8117e29da
Author: Seong-ho Cho <shcho@gnome.org>
Date:	2017-03-19

    Update Korean translation

M	po/ko.po

commit 64f79f8656e06cd0f7e34c1847a84edc385f26e8
Author: Rūdolfs Mazurs <rudolfsm@src.gnome.org>
Date:	2017-03-18

    Update Latvian translation

M	po/lv.po

commit 08e0323bd5b48d3aedeb65211ea0e93f2fa70c7c
Author: gogo <trebelnik2@gmail.com>
Date:	2017-03-15

    Update Croatian translation

M	po/hr.po

commit 635417ba7ae6eefba038e7af7daf8b1d997d611f
Author: Luca Ferretti <lferrett@gnome.org>
Date:	2017-03-15

    Update Italian translation

M	po/it.po

commit 91155a3262280d2ce98541bb69525048392d3896
Author: Milo Casagrande <milo@ubuntu.com>
Date:	2017-03-15

    Update Italian translation

M	po/it.po

commit abf271f7c70312b9e50e8c4edcf31c25ab0b1798
Author: Jordi Mas <jmas@softcatala.org>
Date:	2017-03-12

    Update Catalan translation

M	po/ca.po

commit 546a5c60fd1082e2aa2216929ecce8be315b3105
Author: A S Alam <aalam@users.sf.net>
Date:	2017-03-11

    updated Punjabi Translation

M	po/pa.po

commit 4810b95394db1ed743ef3a73231a52d21d4dde33
Author: Fran Dieguez <frandieguez@gnome.org>
Date:	2017-03-11

    Update Galician translation

M	po/gl.po

commit 6d1c4f2dda48008b85be7e799653216df72d0437
Author: Trần Ngọc Quân <vnwildman@gmail.com>
Date:	2017-03-10

    Updated Vietnamese translation

    Signed-off-by: Trần Ngọc Quân <vnwildman@gmail.com>

M	po/vi.po

commit 8902c9d1004262860cfcd0f6d4baa9ec7138c28e
Author: Aurimas Černius <aurisc4@gmail.com>
Date:	2017-03-04

    Updated Lithuanian translation

M	po/lt.po

commit 28b35a923aba2c38d345f6a1b70b43435ea96323
Author: Rafael Fontenelle <rafaelff@gnome.org>
Date:	2017-03-04

    Update Brazilian Portuguese translation

A	help/pt_BR/figures/add-text-annotation.png
A	help/pt_BR/figures/annotations-nav-to-page.png
A	help/pt_BR/figures/annotations-navigate.png
A	help/pt_BR/figures/list-add-tabs.png
A	help/pt_BR/figures/print-select.png
A	help/pt_BR/figures/reverse-collate.png
A	help/pt_BR/figures/zoom.png
M	help/pt_BR/pt_BR.po

commit 477256d2181a8cf22c6c11ec4e9fbf970e2ce578
Author: Joe Hansen <joedalton2@yahoo.dk>
Date:	2017-02-27

    Updated Danish translation

M	po/da.po

commit e5db68bd87868827dabaabe14a718d2ce919eb6b
Author: Joe Hansen <joedalton2@yahoo.dk>
Date:	2017-02-27

    Updated Danish translation

M	po/da.po

commit ffa8f15c1131147ff50af0275ff43b67409c465c
Author: Мирослав Николић <miroslavnikolic@rocketmail.com>
Date:	2017-02-26

    Updated Serbian translation

M	po/sr.po
M	po/sr@latin.po

commit e4624a08b0c628e1ba8bed7f2cbb08c3a700a0ce
Author: Rafael Fontenelle <rafaelff@gnome.org>
Date:	2017-02-25

    Update Brazilian Portuguese translation

M	po/pt_BR.po

commit 0abb5018e9e4307a4ccc170f4431307a70d7ee6b
Author: Claude Paroz <claude@2xlibre.net>
Date:	2017-02-25

    Updated French translation

M	po/fr.po

commit 83a519c25b54f337a8501027ffcd6866622c2dd3
Author: Chao-Hsiung Liao <j_h_liau@yahoo.com.tw>
Date:	2017-02-22

    Update Chinese (Taiwan) translation

M	po/zh_TW.po

commit 7b31769578c429dd7de204a1d81f37d09ebeb0a8
Author: Daniel Korostil <ted.korostiled@gmail.com>
Date:	2017-02-20

    Updated Ukrainian translation

M	po/uk.po

commit a69b84a7020718ddb61608ec230840db2f2ae66c
Author: Baurzhan Muftakhidinov <baurthefirst@gmail.com>
Date:	2017-02-19

    Update Kazakh translation

M	po/kk.po

commit aa0a5de3f0d8a52c47b7bbeb470164793a1ee2d5
Author: Kukuh Syafaat <syafaatkukuh@gmail.com>
Date:	2017-02-19

    Update Indonesian translation

M	po/id.po

commit 5c31cab4531d36d51fa23e56b38ca40de61efc5f
Author: Ask Hjorth Larsen <asklarsen@gmail.com>
Date:	2017-02-16

    Updated Danish translation

M	po/da.po

commit b2ddf7a1a17663119f24c8facd1fc1bcf712ee93
Author: Jiri Grönroos <jiri.gronroos@iki.fi>
Date:	2017-02-15

    Update Finnish translation

M	po/fi.po

commit 32376612585d5c5d04ec7c3ec03138c349d9f4c3
Author: Kukuh Syafaat <syafaatkukuh@gmail.com>
Date:	2017-02-13

    Update Indonesian translation

M	po/id.po

commit 0f2dacc0ccacb87362182508e6fece0690230290
Author: Inaki Larranaga Murgoitio <dooteo@zundan.com>
Date:	2017-02-11

    Update Basque language

M	po/eu.po

commit bfea980ade96221253edd40afcdcb68efbe5861c
Author: Fabio Tomat <f.t.public@gmail.com>
Date:	2017-02-02

    Update Friulian translation

M	po/fur.po

commit c8bbc8e5ff5a84340dc0d5c9b974e4d8780128c4
Author: Piotr Drąg <piotrdrag@gmail.com>
Date:	2017-01-20

    Update Polish translation

M	po/pl.po

commit dd4d2a991e1e1677f05d1e11fcd65cb8c85e13ff
Author: Dušan Kazik <prescott66@gmail.com>
Date:	2017-01-18

    Update Slovak translation

M	po/sk.po

commit 1761e860aa38ff56a4046d1abb3da6c875dce07c
Author: Dušan Kazik <prescott66@gmail.com>
Date:	2017-01-16

    Update Slovak translation

M	po/sk.po

commit a4b6f7ca00e08f84678e518c904f982b41fe7614
Author: Simon Nagl <simonnagl@aim.com>
Date:	2016-12-02

    Hide search bar when entering presentation mode

    Before fix:
    1. Press Ctrl+f to display search bar
    2. Press F5 to enter presentation mode
    -> Search bar is still there but without feedback on the document.

    https://bugzilla.gnome.org/show_bug.cgi?id=645976 descides:
    "Search is not supposed to work in presentation mode, so the find
    bar shoudn't be displayed."

    This commit hides the search bar before entering presentation mode.

    https://bugzilla.gnome.org//show_bug.cgi?id=775536

M	shell/ev-window.c

commit 932f2d0ca69aa3a75eb6598507f64ca75b96d8d0
Author: Trinh Anh Ngoc <atw1990@gmail.com>
Date:	2017-01-13

    Wrap searchbar in a revealer

    Workaround for the gtk+ bug:
    https://bugzilla.gnome.org/show_bug.cgi?id=724096

    https://bugzilla.gnome.org/show_bug.cgi?id=777204

M	shell/ev-window.c


shell/libsec.c
/*
 *  Copyright (C) 2017 valoq <valoq@mailbox.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 */

#include "libsec.h"
#include <stdio.h>

#define HAVE_LIBSECCOMP
#ifdef HAVE_LIBSECCOMP

#include <seccomp.h> /* libseccomp */
#include <sys/prctl.h> /* prctl */
#include <sys/socket.h>
#include <fcntl.h>
#include <stdlib.h>
#include <errno.h>

#define DENY_RULE(call) { if (seccomp_rule_add (ctx, SCMP_ACT_KILL, SCMP_SYS(call), 0) < 0) goto out; }
#define ALLOW_RULE(call) { if (seccomp_rule_add (ctx, SCMP_ACT_ALLOW, SCMP_SYS(call), 0) < 0) goto out; }

scmp_filter_ctx ctx;

int protectedMode(void){

    // prevent child processes from getting more priv e.g. via setuid, capabilities, ...
    //prctl(PR_SET_NO_NEW_PRIVS, 1);

    if (prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0)) {
        perror("prctl SET_NO_NEW_PRIVS");
        exit(EXIT_FAILURE);
    }


    // prevent escape via ptrace
    //prctl(PR_SET_DUMPABLE, 0);

    if(prctl (PR_SET_DUMPABLE, 0, 0, 0, 0)){
        perror("prctl PR_SET_DUMPABLE");
        exit(EXIT_FAILURE);
    }


    // initialize the filter
    ctx = seccomp_init(SCMP_ACT_ALLOW);
    if (ctx == NULL)
        return 1;

    DENY_RULE (_sysctl);
    DENY_RULE (acct);
    DENY_RULE (add_key);
    DENY_RULE (adjtimex);
    DENY_RULE (chroot);
    DENY_RULE (clock_adjtime);
    DENY_RULE (create_module);
    DENY_RULE (delete_module);
    DENY_RULE (fanotify_init);
    DENY_RULE (finit_module);
    DENY_RULE (get_kernel_syms);
    DENY_RULE (get_mempolicy);
    DENY_RULE (init_module);
    DENY_RULE (io_cancel);
    DENY_RULE (io_destroy);
    DENY_RULE (io_getevents);
    DENY_RULE (io_setup);
    DENY_RULE (io_submit);
    DENY_RULE (ioperm);
    DENY_RULE (iopl);
    DENY_RULE (ioprio_set);
    DENY_RULE (kcmp);
    DENY_RULE (kexec_file_load);
    DENY_RULE (kexec_load);
    DENY_RULE (keyctl);
    DENY_RULE (lookup_dcookie);
    DENY_RULE (mbind);
    DENY_RULE (nfsservctl);
    DENY_RULE (migrate_pages);
    DENY_RULE (modify_ldt);
    DENY_RULE (mount);
    DENY_RULE (move_pages);
    DENY_RULE (name_to_handle_at);
    DENY_RULE (open_by_handle_at);
    DENY_RULE (perf_event_open);
    DENY_RULE (pivot_root);
    DENY_RULE (process_vm_readv);
    DENY_RULE (process_vm_writev);
    DENY_RULE (ptrace);
    DENY_RULE (reboot);
    DENY_RULE (remap_file_pages);
    DENY_RULE (request_key);
    DENY_RULE (set_mempolicy);
    DENY_RULE (swapoff);
    DENY_RULE (swapon);
    DENY_RULE (sysfs);
    DENY_RULE (syslog);
    DENY_RULE (tuxcall);
    DENY_RULE (umount2);
    DENY_RULE (uselib);
    DENY_RULE (vmsplice);


    //applying filter...
    if (seccomp_load (ctx) >= 0){
	// free ctx after the filter has been loaded into the kernel
	seccomp_release(ctx);
        return 0;
    }

    
  out:
    //something went wrong
    seccomp_release(ctx);
    return 1;
}


int protectedView(void){

    // prevent child processes from getting more priv e.g. via setuid, capabilities, ...
    //prctl(PR_SET_NO_NEW_PRIVS, 1);

    if (prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0)) {
        perror("prctl SET_NO_NEW_PRIVS");
        exit(EXIT_FAILURE);
    }


    // prevent escape via ptrace
    //prctl(PR_SET_DUMPABLE, 0);

    if(prctl (PR_SET_DUMPABLE, 0, 0, 0, 0)){
        perror("prctl PR_SET_DUMPABLE");
        exit(EXIT_FAILURE);
    }

    // initialize the filter
    ctx = seccomp_init(SCMP_ACT_KILL);
    if (ctx == NULL)
        return 1;


    ALLOW_RULE (access);
    ALLOW_RULE (arch_prctl);
    ALLOW_RULE (brk);
    ALLOW_RULE (chdir);
    ALLOW_RULE (chmod);
    ALLOW_RULE (clock_getres);
    ALLOW_RULE (clone);
    ALLOW_RULE (close);
    ALLOW_RULE (connect);
      ALLOW_RULE (dup2);
    ALLOW_RULE (eventfd2);
    ALLOW_RULE (exit);
    ALLOW_RULE (exit_group);
    ALLOW_RULE (fadvise64);
    ALLOW_RULE (fallocate);
    ALLOW_RULE (fcntl);
    ALLOW_RULE (fstat);
    ALLOW_RULE (fstatfs);
    ALLOW_RULE (fsync);
    ALLOW_RULE (futex);
    ALLOW_RULE (getdents);
    ALLOW_RULE (getegid);
    ALLOW_RULE (geteuid);
      ALLOW_RULE (getgid);
      ALLOW_RULE (getuid);
    ALLOW_RULE (getpid);
      ALLOW_RULE (getppid);
    ALLOW_RULE (getpeername);
      ALLOW_RULE (getpgrp);
      ALLOW_RULE (getrandom);
    ALLOW_RULE (getresgid);
    ALLOW_RULE (getresuid);
    ALLOW_RULE (getrlimit);
    ALLOW_RULE (getsockname);
    ALLOW_RULE (getsockopt);  /* needed for access to x11 socket in network namespace (without abstract sockets) */
    ALLOW_RULE (getuid);
    ALLOW_RULE (inotify_add_watch);
    ALLOW_RULE (inotify_init1);
    ALLOW_RULE (inotify_rm_watch);
    ALLOW_RULE (ioctl);
    ALLOW_RULE (lseek);
    ALLOW_RULE (lstat);
    ALLOW_RULE (madvise);
    ALLOW_RULE (mmap);
    ALLOW_RULE (mprotect);
    ALLOW_RULE (munmap);
    //ALLOW_RULE (open);
      ALLOW_RULE (pipe);
    ALLOW_RULE (poll);
    ALLOW_RULE (prctl);
    ALLOW_RULE (pread64);
    ALLOW_RULE (prlimit64);
    ALLOW_RULE (pwrite64);
    ALLOW_RULE (read);
    ALLOW_RULE (recvfrom);
    ALLOW_RULE (recvmsg);
    ALLOW_RULE (rename);
    ALLOW_RULE (restart_syscall);
    ALLOW_RULE (rt_sigaction);
    ALLOW_RULE (rt_sigprocmask);
    ALLOW_RULE (sendmsg);
    ALLOW_RULE (sendto);
    ALLOW_RULE (set_robust_list);
    ALLOW_RULE (set_tid_address);
    ALLOW_RULE (shmat);
    ALLOW_RULE (shmctl);
    ALLOW_RULE (shmdt);
    ALLOW_RULE (shmget);
    ALLOW_RULE (shutdown);
    //ALLOW_RULE (socket);
    ALLOW_RULE (stat);
    ALLOW_RULE (statfs);
      ALLOW_RULE (sysinfo);
    ALLOW_RULE (uname);
      ALLOW_RULE (wait4);
    ALLOW_RULE (write);
    ALLOW_RULE (writev);

    //link
    //unlink

    // needed for save copy, (writing new files) -- to be reviewed

    ALLOW_RULE (flistxattr);
    ALLOW_RULE (getcwd);
    ALLOW_RULE (lchown);
    ALLOW_RULE (listxattr);
    ALLOW_RULE (mkdir);
    ALLOW_RULE (pipe2);
    ALLOW_RULE (readlink);
    ALLOW_RULE (rmdir);
    ALLOW_RULE (splice);

    ALLOW_RULE (unlink);
    ALLOW_RULE (utimes);


    // prevent network access and open files for writing

    /* special restrictions for socket, only allow AF_UNIX/AF_LOCAL */
    if (seccomp_rule_add (ctx, SCMP_ACT_ALLOW, SCMP_SYS(socket), 1,
			  SCMP_CMP(0, SCMP_CMP_EQ, AF_UNIX)) < 0)
	goto out;

    if (seccomp_rule_add (ctx, SCMP_ACT_ALLOW, SCMP_SYS(socket), 1,
			  SCMP_CMP(0, SCMP_CMP_EQ, AF_LOCAL)) < 0)
	goto out;


    /* special restrictions for open, allow only reading files */
    if (seccomp_rule_add (ctx, SCMP_ACT_ALLOW, SCMP_SYS(open), 1,
			  SCMP_CMP(1, SCMP_CMP_MASKED_EQ, O_WRONLY | O_RDWR, 0)) < 0)
	goto out;

    if (seccomp_rule_add (ctx, SCMP_ACT_ERRNO (EACCES), SCMP_SYS(open), 1,
			  SCMP_CMP(1, SCMP_CMP_MASKED_EQ, O_WRONLY, O_WRONLY)) < 0)
	goto out;

    if (seccomp_rule_add (ctx, SCMP_ACT_ERRNO (EACCES), SCMP_SYS(open), 1,
			  SCMP_CMP(1, SCMP_CMP_MASKED_EQ, O_RDWR, O_RDWR)) < 0)
	goto out;

	
    /* restrict fcntl calls */
    // this syscall sets the file descriptor to read write
    // if (seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(fcntl), 1,
    //                             SCMP_CMP(0, SCMP_CMP_EQ, 3)) < 0 )
    //     goto out;
    // fcntl(3, F_GETFL)                       = 0x2 (flags O_RDWR)
    // fcntl(3, F_SETFL, O_RDWR|O_NONBLOCK)    = 0
    // fcntl(3, F_SETFD, FD_CLOEXEC)           = 0


    //applying filter...
    if (seccomp_load (ctx) >= 0){
	// free ctx after the filter has been loaded into the kernel
	seccomp_release(ctx);
        return 0;
    }

	
 out:
    //something went wrong
    seccomp_release(ctx);
    return 1;
}

#else /* HAVE_LIBSECCOMP */


int protectedMode(void){

    perror("No seccomp support compiled-in\n");
    return 1;
}

int protectedView(void){

    perror("No seccomp support compiled-in\n");
    return 1;
}

#endif /* HAVE_LIBSECCOMP */


shell/libsec.h
/*
 *  Copyright (C) 2017 valoq <valoq@mailbox.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 */

#ifndef SECCOMP_H
#define SECCOMP_H

/* basic filter */
// this mode allows normal use
// only dangerous syscalls are blacklisted
int protectedMode(void);

/* secure read-only mode */
// whitelist minimal syscalls only
// this mode does not allow writing files
// or to open external links and applications
// network connections are prohibited as well
int protectedView(void);

#endif


shell/libsecd.h
/*
 *  Copyright (C) 2017 valoq <valoq@mailbox.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 */

#ifndef SECCOMP_H
#define SECCOMP_H

// whitelist minimal syscalls only
int applyFilter(void);

#endif


shell/main.c
/*
 *  Copyright (C) 2004 Marco Pesenti Gritti
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 */

#include "config.h"

//todo: set this flag in makefile
#define HAVE_LIBSECCOMP
//#define USE_PROTECTEDVIEW

#ifdef USE_PROTECTEDVIEW
#ifndef HAVE_LIBSECCOMP
#define HAVE_LIBSECCOMP
#endif /* HAVE_LIBSECCOMP */
#endif /* USE_PROTECTEDVIEW */


#ifdef HAVE_LIBSECCOMP
#include "libsec.h"
#endif /* HAVE_LIBSECCOMP */


#include <stdlib.h>
#include <string.h>

#include <glib/gstdio.h>
#include <glib/gi18n.h>
#include <gtk/gtk.h>

#include "ev-application.h"
#include "ev-debug.h"
#include "ev-init.h"
#include "ev-file-helpers.h"
#include "ev-stock-icons.h"
#include "ev-metadata.h"

#ifdef G_OS_WIN32
#include <io.h>
#include <conio.h>
#if !(_WIN32_WINNT >= 0x0500)
#error "_WIN32_WINNT must be defined >= 0x0500"
#endif
#include <windows.h>
#endif

static gchar   *ev_page_label;
static gchar   *ev_find_string;
static gint     ev_page_index = 0;
static gchar   *ev_named_dest;
static gboolean preview_mode = FALSE;
static gboolean fullscreen_mode = FALSE;
static gboolean presentation_mode = FALSE;
static gboolean unlink_temp_file = FALSE;
static gchar   *print_settings;
static const char **file_arguments = NULL;


static gboolean
option_version_cb (const gchar *option_name,
                   const gchar *value,
                   gpointer     data,
                   GError     **error)
{
  g_print ("%s %s\n", _("GNOME Document Viewer"), VERSION);

  exit (0);
  return FALSE;
}

static const GOptionEntry goption_options[] =
{
	{ "page-label", 'p', 0, G_OPTION_ARG_STRING, &ev_page_label, N_("The page label of the document to display."), N_("PAGE")},
	{ "page-index", 'i', 0, G_OPTION_ARG_INT, &ev_page_index, N_("The page number of the document to display."), N_("NUMBER")},
	{ "named-dest", 'n', 0, G_OPTION_ARG_STRING, &ev_named_dest, N_("Named destination to display."), N_("DEST")},
	{ "fullscreen", 'f', 0, G_OPTION_ARG_NONE, &fullscreen_mode, N_("Run evince in fullscreen mode"), NULL },
	{ "presentation", 's', 0, G_OPTION_ARG_NONE, &presentation_mode, N_("Run evince in presentation mode"), NULL },
	{ "preview", 'w', 0, G_OPTION_ARG_NONE, &preview_mode, N_("Run evince as a previewer"), NULL },
	{ "find", 'l', 0, G_OPTION_ARG_STRING, &ev_find_string, N_("The word or phrase to find in the document"), N_("STRING")},
	{ "unlink-tempfile", 'u', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE, &unlink_temp_file, NULL, NULL },
	{ "print-settings", 't', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_FILENAME, &print_settings, NULL, NULL },
	{ "version", 0, G_OPTION_FLAG_NO_ARG | G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_CALLBACK, option_version_cb, NULL, NULL },
	{ G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &file_arguments, NULL, N_("[FILE…]") },
	{ NULL }
};

static gboolean
launch_previewer (void)
{
	GString *cmd_str;
	gchar   *cmd;
	gboolean retval = FALSE;
	GError  *error = NULL;

	/* Rebuild the command line, ignoring options
	 * not supported by the previewer and taking only
	 * the first path given
	 */
	cmd_str = g_string_new ("evince-previewer");
		
	if (print_settings) {
		gchar *quoted;

		quoted = g_shell_quote (print_settings);
		g_string_append_printf (cmd_str, " --print-settings %s", quoted);
		g_free (quoted);
	}

	if (unlink_temp_file)
		g_string_append (cmd_str, " --unlink-tempfile");

	if (file_arguments) {
		gchar *quoted;
		
		quoted = g_shell_quote (file_arguments[0]);
		g_string_append_printf (cmd_str, " %s", quoted);
		g_free (quoted);
	}

	cmd = g_string_free (cmd_str, FALSE);

	if (!error) {
		GAppInfo *app;

		app = g_app_info_create_from_commandline (cmd, NULL, 0, &error);

		if (app != NULL) {
			retval = g_app_info_launch (app, NULL, NULL, &error);
			g_object_unref (app);
		}
	}

	if (error) {
		g_warning ("Error launching previewer: %s\n", error->message);
		g_error_free (error);
	}

	g_free (cmd);

	return retval;
}

static gchar *
get_label_from_filename (const gchar *filename)
{
	GFile   *file;
	gchar   *label;
	gboolean exists;

	label = g_strrstr (filename, "#");
	if (!label)
		return NULL;

	/* Filename contains a #, check
	 * whether it's part of the path
	 * or a label
	 */
	file = g_file_new_for_commandline_arg (filename);
	exists = g_file_query_exists (file, NULL);
	g_object_unref (file);

	return exists ? NULL : label;
}

static void
load_files (const char **files)
{
	GdkScreen       *screen = gdk_screen_get_default ();
	EvWindowRunMode  mode = EV_WINDOW_MODE_NORMAL;
	gint             i;
	EvLinkDest      *global_dest = NULL;

	if (!files) {
		if (!ev_application_has_window (EV_APP))
			ev_application_open_recent_view (EV_APP, screen, GDK_CURRENT_TIME);
		return;
	}

	if (ev_page_label)
		global_dest = ev_link_dest_new_page_label (ev_page_label);
	else if (ev_page_index)
		global_dest = ev_link_dest_new_page (MAX (0, ev_page_index - 1));
	else if (ev_named_dest)
		global_dest = ev_link_dest_new_named (ev_named_dest);

	if (fullscreen_mode)
		mode = EV_WINDOW_MODE_FULLSCREEN;
	else if (presentation_mode)
		mode = EV_WINDOW_MODE_PRESENTATION;

	for (i = 0; files[i]; i++) {
		const gchar *filename;
		gchar       *uri;
		gchar       *label;
		GFile       *file;
		EvLinkDest  *dest = NULL;
		const gchar *app_uri;

		filename = files[i];
		label = get_label_from_filename (filename);
		if (label) {
			*label = 0;
			label++;
			dest = ev_link_dest_new_page_label (label);
		} else if (global_dest) {
			dest = g_object_ref (global_dest);
		}

		file = g_file_new_for_commandline_arg (filename);
		uri = g_file_get_uri (file);
		g_object_unref (file);

		app_uri = ev_application_get_uri (EV_APP);
		if (app_uri && strcmp (app_uri, uri) == 0) {
			g_free (uri);
			continue;
		}



		ev_application_open_uri_at_dest (EV_APP, uri, screen, dest,
						 mode, ev_find_string,
						 GDK_CURRENT_TIME);

		if (dest)
			g_object_unref (dest);
		g_free (uri);
        }
}

int
main (int argc, char *argv[])
{

#ifdef HAVE_LIBSECCOMP

        if(protectedMode()){
	    perror("SECCOMP initialisation failed");
	    exit(EXIT_FAILURE);
	}

#endif /* HAVE_LIBSECCOMP */


#ifdef USE_PROTECTEDVIEW

	if(protectedView()){
	    perror("SECCOMP initialisation failed");
	    exit(EXIT_FAILURE);
	}
	
#endif /* USE_PROTECTEDVIEW */


        EvApplication  *application;
	GOptionContext *context;
	GError         *error = NULL;
        int             status;

#ifdef G_OS_WIN32

    if (fileno (stdout) != -1 &&
 	  _get_osfhandle (fileno (stdout)) != -1)
	{
	  /* stdout is fine, presumably redirected to a file or pipe */
	}
    else
    {
	  typedef BOOL (* WINAPI AttachConsole_t) (DWORD);

	  AttachConsole_t p_AttachConsole =
	    (AttachConsole_t) GetProcAddress (GetModuleHandle ("kernel32.dll"), "AttachConsole");

	  if (p_AttachConsole != NULL && p_AttachConsole (ATTACH_PARENT_PROCESS))
      {
	      freopen ("CONOUT$", "w", stdout);
	      dup2 (fileno (stdout), 1);
	      freopen ("CONOUT$", "w", stderr);
	      dup2 (fileno (stderr), 2);

      }
	}
#endif

#ifdef ENABLE_NLS
	/* Initialize the i18n stuff */
	bindtextdomain (GETTEXT_PACKAGE, ev_get_locale_dir());
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);
#endif

	context = g_option_context_new (N_("GNOME Document Viewer"));
	g_option_context_set_translation_domain(context, GETTEXT_PACKAGE);
	g_option_context_add_main_entries (context, goption_options, GETTEXT_PACKAGE);

	g_option_context_add_group (context, gtk_get_option_group (TRUE));

	if (!g_option_context_parse (context, &argc, &argv, &error)) {
		g_printerr ("Cannot parse arguments: %s\n", error->message);
		g_error_free (error);
		g_option_context_free (context);

		return 1;
	}
	g_option_context_free (context);

	if (preview_mode) {
		gboolean retval;
		
		retval = launch_previewer ();
		
		return retval ? 0 : 1;
	}

        if (!ev_init ())
                return 1;

	ev_stock_icons_init ();

	/* Manually set name and icon */
	g_set_application_name (_("Document Viewer"));
	gtk_window_set_default_icon_name ("evince");

        application = ev_application_new ();
        if (!g_application_register (G_APPLICATION (application), NULL, &error)) {
                g_printerr ("Failed to register: %s\n", error->message);
                g_error_free (error);
                status = 1;
                goto done;
        }


	// enhance syscall filter at this point, only few syscalls should be needed

	load_files (file_arguments);

	/* Change directory so we don't prevent unmounting in case the initial cwd
	 * is on an external device (see bug #575436)
	 */
	g_chdir (g_get_home_dir ());

	status = g_application_run (G_APPLICATION (application), 0, NULL);

    done:
	ev_shutdown ();
	ev_stock_icons_shutdown ();

        g_object_unref (application);
	return status;
}
